src/NextGen/DonationForm/DataTransferObjects/DonateFormRouteData.php

```diff
@@ -0,0 +1,150 @@
+        $self->requestData = $requestData;
+
+        return $self;
+    }
+
+    /**
+     * This method loops over the form schema to
+     * compares the request against the individual fields,
+     * their types and validation rules.
+     *
+     * @unreleased
+     *
+     * @throws DonationFormFieldErrorsException
+     */
+    public function validated(): DonateControllerData
+    {
+        $request = $this->getRequestData();
+        $validData = new DonateControllerData();
+
+        $errors = [];
+
+        /** @var DonationForm $form */
+        $form = DonationForm::find($this->formId);
+
+        $form->schema()->walkFields(function (Field $field) use (
+            $request,
+            $validData,
+            &$errors
+        ) {
+            $fieldValue = $request[$field->getName()];
+
+            // validate specific field types like email
+            if ($field->getType() === Email::TYPE && !is_email($fieldValue)) {
+                $error = $request[$field->getName()] = [
+                    'error_id' => $field->getName(),
+                    'error_message' => sprintf(
+                        esc_html__('Please use a valid email for %s', 'give'),
+                        $field->getName()
+                    ),
+                ];
+
+                $errors[] = $error;
+            }
+
+            // validate required fields
+            if (empty($fieldValue) && $field->isRequired()) {
+                $errors[] = $request[$field->getName()] = $field->getRequiredError();
+            }
+
+            if (empty($errors[$field->getName()])) {
+                $validData->{$field->getName()} = $this->castFieldValue($field, $fieldValue ?? '');
+            }
+        });
+
+        if ($errors) {
+            $this->throwDonationFormFieldErrorsException($errors);
+        }
+
+        $validData->wpUserId = get_current_user_id();
+
+        return $validData;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function getRequestData(): array
+    {
+        return $this->requestData;
+    }
+
+    /**
+     * This loops over an array of errors in the specific FieldAPI format,
+     * and converts them into a WP_Error object that is attached to the
+     * exception and delivered back to the client via JSON.
+     *
+     * @unreleased
+     *
+     * @param  array{error_id: string, error_message: string}  $errors
+     *
+     * @throws DonationFormFieldErrorsException
+     */
+    private function throwDonationFormFieldErrorsException(array $errors)
+    {
+        $wpError = new WP_Error();
+
+        foreach ($errors as $error) {
+            $wpError->add($error['error_id'], $error['error_message']);
+        }
+
+        $exception = new DonationFormFieldErrorsException();
+        $exception->setError($wpError);
+
+        throw $exception;
+    }
+
+    /**
+     * Some properties need to be cast to specific types.
+     *
+     * TODO: figure out a less static way of doing this
```

Comment:
_Later_, we can use the validation rules — "boolean", "string", "int", and so forth — to provide casting hints.